File: 0806-domino-and-tromino-tiling/0806-domino-and-tromino-tiling.py
class Solution:
    def numTilings (self, n : int) -> int:
        MOD = 10 ** 9 + 7
        if n == 0:
            return 1
        if n == 1:
            return 1
        if n == 2:
            return 2
        dp = [0] * (n + 1)
        prefix = [0] * (n + 1)
        dp [0] = 1
        dp [1] = 1
        dp [2] = 2
        prefix [0] = 1
        prefix [1] = 2
        prefix [2] = 4
        for i in range (3, n + 1):
            dp [i] = (dp [i - 1] + dp [i - 2] + 2 * prefix [i - 3]) % MOD
            prefix [i] = (prefix [i - 1] + dp [i]) % MOD
        return dp [n]

File: 0806-domino-and-tromino-tiling/0806-domino-and-tromino-tiling.java
class Solution {
    ArrayList <Integer> dp = new ArrayList <> (Collections.nCopies (1001, -1)); 
    final int MOD = 1_000_000_007;
    public int solve (int n) {
        if (n == 0) return 0;
        if (n == 1) return 1;
        if (n == 2) return 2;
        if (n == 3) return 5;
        if (dp.get (n) != -1) return dp.get (n);
        long part1 = (2L * solve (n - 1)) % MOD;
        long part2 = solve (n - 3) % MOD;
        int result = (int) ((part1 + part2) % MOD);
        dp.set (n, result);
        return result;
    }
    public int numTilings (int n) {
        return solve (n);
    }
}